### ISSUE TITLE
Question: Programatically creating a refresh token

### ISSUE THREAD

**Issue by joesaunderson** [markitosgv/JWTRefreshTokenBundle#270]
Is it possible to programatically generate a refresh token? Similar to https://github.com/lexik/LexikJWTAuthenticationBundle/blob/2.x/Resources/doc/7-manual-token-creation.md

I want to handle the login / refresh option in GraphQL, so I want to generate the two tokens myself in a PHP resolver.

Or is there a better way of doing this?

**Comment by joesaunderson** [comment]
Think I have solved this with:

```
$refreshToken = $this->refreshTokenManager->createForUserWithTtl($user, 604800);
```

My next question, how do I implement a "manual" refresh in a resolver, and return the refreshed JWT

**Comment by fd6130** [comment]
> Think I have solved this with:
> 
> ```
> $refreshToken = $this->refreshTokenManager->createForUserWithTtl($user, 604800);
> ```
> 
> My next question, how do I implement a "manual" refresh in a resolver, and return the refreshed JWT

Perhaps you mean call a `refresh token` api and pass it the refresh token, then the controller return a refreshed JWT?

I think you can do that by creating your own controller and logic.

**Comment by joesaunderson** [comment]
> Not sure what do you mean by resolver, do you mean Symfony resolver or something else?

I am using GraphQL, which has the idea of Resolvers (essentially a Symfony controller). So my question is, how do I programatically do what the `/api/refresh` API was doing behind the scenes, if that makes sense?

**Comment by fd6130** [comment]
Haven't learn about GraphQL yet. But for the `/api/refresh`, the logic is quite simple. You pass in the refresh token, check if the refresh token is exist in database and if it is still in fresh, generate a new jwt token and return it to the user.

For example in Symfony Controller:

```php

/**
* @Route("/api/refresh")
*/
public function refreshJWT(Request $request)
{
     // check the database for the refresh token.
    // if exist and is still fresh, generate a new jwt token and return it.
   $jwt = // the generate jwt code ;
   return $this->json(['token' => $jwt]);
}
```

**Comment by joesaunderson** [comment]
Thanks for confirming that, makes sense to me now!

**Comment by fd6130** [comment]
> Thanks for confirming that, makes sense to me now!

It's been a while from this issue, I think it can be close now?

**Comment by saifgo** [comment]
I tried the solution provided by @joesaunderson 
but it didn't work and had to change it to this 
```
use Gesdinet\JWTRefreshTokenBundle\Generator\RefreshTokenGeneratorInterface;
use Gesdinet\JWTRefreshTokenBundle\Model\RefreshTokenManagerInterface;
```
```
private $RefreshTokenGenerator;
public function __construct(
        RefreshTokenGeneratorInterface $refreshTokenGenerator,
        RefreshTokenManagerInterface $refreshTokenManager) 
    {
        $this->refreshTokenGenerator = $refreshTokenGenerator;
        $this->refreshTokenManager = $refreshTokenManager;
    }
```
and when i want to create refresh token i do this

```
 $refreshToken = $this->refreshTokenGenerator->createForUserWithTtl($user, 604800);
 $this->refreshTokenManager->save($refreshToken);
 $refreshTokenString = $refreshToken->getRefreshToken();
```
after doing this the $refreshTokenString will be the refresh token and the refreshToken will be saved in the database for future use  as @fd6130  cleared

**Comment by MedDimessi** [comment]
@fd6130 Can you please provider the 'use directory for the `refreshTokenManager->createForUserWithTtl($user, 604800);`,
where does the function `createForUserWithTtl()` exist??
I'm using Symfony 5.3.

**Comment by saifgo** [comment]
@MedDimessi if you follow the code on my response you will know it cames from "refreshTokenGenerator" that you can use it with "Gesdinet\JWTRefreshTokenBundle\Generator\RefreshTokenGeneratorInterface;"

**Comment by MedDimessi** [comment]
@saifgo  I did, but the `JWTRefreshTokenBundle:0.12.0` deesn't have the `Generator` folder.
That why I specified what version of Symfony I'm using.

>  I'm using Symfony 5.3.

**Comment by orangevinz** [comment]
@MedDimessi Same for me.. Did you find a solution ?

> @fd6130 Can you please provider the 'use directory for the `refreshTokenManager->createForUserWithTtl($user, 604800);`, where does the function `createForUserWithTtl()` exist?? I'm using Symfony 5.3.

**Comment by orangevinz** [comment]
Ok I was not on the version `1.0`  my bad

**Comment by atanasov** [comment]
If someone still needs this working on Symfony  5.3.* and "gesdinet/jwt-refresh-token-bundle": "^0.12" 

`
use Gesdinet\JWTRefreshTokenBundle\Model\RefreshTokenManagerInterface;
use Symfony\Component\PropertyAccess\PropertyAccessor;
`

`
    private function createRefreshToken(UserInterface $user)
    {
        $ttl = 2592000;
        $userIdentityField = 'username';
        $datetime = new \DateTime();
        $datetime->modify('+'.$ttl.' seconds');

        $refreshToken = $this->refreshTokenManager->create();

        $accessor = new PropertyAccessor();
        $userIdentityFieldValue = $accessor->getValue($user, $userIdentityField);

        $refreshToken->setUsername($userIdentityFieldValue);
        $refreshToken->setRefreshToken();
        $refreshToken->setValid($datetime);

        $this->refreshTokenManager->save($refreshToken);
        return $refreshToken->getRefreshToken();
    }
`

**Comment by shakaran** [comment]
@joesaunderson, yes, and it is documented now — bc7a96a, prompted by #322 which asked the same thing.

**First, the snippets in this thread do not work on current versions**, so anyone arriving here from a search should skip them:

- `$this->refreshTokenManager->createForUserWithTtl(...)` — that method is on `RefreshTokenGeneratorInterface`, not the manager, and on its own it creates a token without storing it. Whatever it did in 0.12, it will not resolve today.
- @atanasov's version uses `$manager->create()` and `setRefreshToken()` with no argument. Neither exists any more; `setRefreshToken(string $refreshToken)` takes the value.

**Creating one.** In a controller or resolver, dispatch the event Lexik dispatches, and the bundle's listener applies your whole configuration — ttl, `single_use`, the cookie and its settings, `return_expiration`:

```php
$data = ['token' => $this->jwtManager->create($user)];
$response = new JsonResponse($data);

$event = new AuthenticationSuccessEvent($data, $user, $response);
$this->eventDispatcher->dispatch($event, Events::AUTHENTICATION_SUCCESS);

$response->setData($event->getData());
```

Using the generator and manager directly, as the thread suggests, gets you a token but none of that configuration — which is fine if you only want the string, and not what you want if you have `single_use` or cookies turned on.

**Your second question — a manual refresh in a resolver** — went unanswered, and @fd6130's sketch is the right shape but leaves out what makes a refresh a refresh: rotating the token under `single_use`, extending it under `ttl_update`, and rejecting one that has expired.

The listener does all of that already, and it does it by reading the *incoming* refresh token out of the request. Which is the catch for GraphQL: the shipped extractors look for `refresh_token` in the request body, the query string and cookies, and in a GraphQL request it is nested inside `variables`, where none of them will find it. So the listener issues a fresh token and never retires the old one.

A [custom extractor](https://github.com/markitosgv/JWTRefreshTokenBundle#token-extractor) that digs it out of the GraphQL variables is what closes that gap — implement `ExtractorInterface`, and it is picked up automatically. With that in place, dispatching the event from your resolver behaves exactly like the refresh endpoint.

You still validate the incoming token yourself before any of this — `$manager->get($token)`, check `isValid()`, load the user — since the resolver is not behind the firewall authenticator.

Closing this. Sorry the second half of your question sat for four years.
